File: scala/116.scala
object Solution {
    def connect(root: Node): Node = {
        def c(root: Node, right: Node) {
          if (root != null) {
            root.next = right
            c(root.left, root.right)
            c(root.right, if (right == null) null else right.left)
          }
        }
        c(root, null)
        root
    }
}

File: leetcodeJava/116.java
package leetcodeJava;

public Node connect(Node root) {
        if (root == null) return null;
        Node l = root.left, r = root.right;
        while (l != null) {
            l.next = r;
            l = l.right;
            r = r.left;
        }
        connect(root.left);
        connect(root.right);
        return root;
    }
